Code/Light.GuardClauses/Range.cs

```diff
@@ -214,6 +214,57 @@ public static class Range
     /// <param name="value">The value that indicates the exclusive lower boundary of the resulting range.</param>
     /// <returns>A value you can use to fluently define the upper boundary of a new range.</returns>
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static Range<T>.RangeFromInfo FromExclusive<T>(T value) where T : IComparable<T> =>
-        new Range<T>.RangeFromInfo(value, false); 
+    public static Range<T>.RangeFromInfo FromExclusive<T>(T value) where T : IComparable<T> => new (value, false);
+
+    /// <summary>
+    /// Creates a range for the specified enumerable that encompasses all valid indexes.
+    /// </summary>
+    /// <param name="enumerable">
+    /// The count of this enumerable will be used to create the index range. Please ensure that this enumerable
+    /// is actually a collection, not a lazy enumerable.
+    /// </param>
+    /// <exception cref="ArgumentNullException">Thrown when <paramref name="enumerable"/> is null.</exception>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Range<int> For(IEnumerable enumerable)  =>
```

Comment:
I'm somewhat concerned that `Range.For` will count by enumerating if I pass an `ICollection<T>` or `IReadOnlyCollection<T>` that is not an `ICollection`. This is not the performance characteristic I would expect.

Most types in the BCL that implement `ICollection<T>` or `IReadOnlyCollection<T>` *do* also implement `ICollection`, but not all of them (`ArraySegment<T>` being a notable exception). For user types, all bets are off.

Could this be addressed by an additional generic overload `Range<int> For<T>(IEnumerable<T> enumerable)` and checking for `ICollection<T>` and `IReadOnlyCollection<T>` there?

Reply:
`ArraySegment<T>` is a struct, so passing an instance of it to a method accepting an `IEnumerable<T>` would box it. The only way to properly support collections that are structs only implementing the generic collection interfaces is to use generics, like so:

```csharp
public static Range<int> For<TCollection, TItem>(this TCollection collection)
    where TCollection : IReadOnlyCollection<T>;
```

This has drawbacks. The first question that arises is which collection interface does the target class implement? `ICollection<T>` or `IReadOnlyCollection<T>`? Both of them? You cannot support both of them because the method parameter list will be considered equal by the compiler, producing an error:

![Cannot overload based on generic constraint](https://github.com/feO2x/Light.GuardClauses/assets/7884158/b4c39dac-6651-435c-9b01-66f7103f98d7)

Secondly, calling this method is a pain, because you need to resolve all generic parameters:

```csharp
public static void Foo()
{
    var array = new ArraySegment<int>(new[] { 1, 2, 3, 4, 5 });
    var range = Range.For<ArraySegment<int>, int>(array);
}
```

I've been down this rabid hole before and simply decided to go with the already existing `Count` method which checks for `ICollection` and `string` - the BCL is covered except for `ArraySegment<T>` - but usually people just use `Span<T>` and `Memory<T>` nowadays. I can add overloads that accept an `ICollection<T>` or an `IReadOnlyCollection<T>` to obtain the count. The thing is, we should decide on one of them, otherwise the overload cannot be resolved automatically:

![Unable to resolve overload when there are overloads for ICollection<T> and IReadOnlyCollection<T>](https://github.com/feO2x/Light.GuardClauses/assets/7884158/d5f3ff35-c9a6-4cc0-ab4d-43b56068fc66)

The thing is: the interfaces for collections in .NET are not designed that well. We need to make compromises, mine was to take the simple route that you can see with the non-generic  `Count` method.

Reply:
Thank you for elaborating. I agree with most of your points. Working with collections in .NET always involves some tradeoffs. I'm fine with the current implementation if that is the tradeoff you want to make.

Nevertheless I wanted to add the following insights, even if further discussion is moot at this point.

I tried to solve the `ICollection<T>`/`IReadOnlyCollection<T>` dilemma in my original API proposal by providing explicit overloads for common classes (`List<T>`, `T[]`) that implement both these interfaces to resolve the ambiguous invocations. This of course doesn't resolve the ambiguity in all cases, especially when considering user defined collection classes.

`ArraySegment<T>` has been superseded by `Span<T>` and `Memory<T>` to some extent, but still serves a purpose for the folks stuck with .NET Framework/.NET Standard 2.0 in legacy projects. Yes, there is the `System.Memory` NuGet package that provide `Span<T>` and `Memory<T>`. But without `Span<T>` overloads for `Stream` and others, their use is limited.

I don't understand why boxing the `ArraySegment<T>` is a concern specific to `IEnumerable<T>` – doesn't the same hold true for the current implementation that takes an `IEnumerable`?

Reply:
You are correct, the `IEnumerable` overload will also box - the point I didn't really make in my previous response is that if you use performance-oriented collections that are structs, you should probably not call `Range.For` but instead create the range in a different way.

For all collections that are reference types, we could also introduce an overload for `IEnumerable<T>`. My initial thought was that we could just simply call the `Count(this IEnumerable<TSource> source)` extension method of LINQ, but looking at the .NET 8 implementation, they do not cover `IReadOnlyList<T>` in their implementation, either:

```csharp
public static int Count<TSource>(this IEnumerable<TSource> source)
{
    if (source == null)
    {
        ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
    }

    if (source is ICollection<TSource> collectionoft)
    {
        return collectionoft.Count;
    }

    if (source is IIListProvider<TSource> listProv)
    {
        return listProv.GetCount(onlyIfCheap: false);
    }

    if (source is ICollection collection)
    {
        return collection.Count;
    }

    int count = 0;
    using (IEnumerator<TSource> e = source.GetEnumerator())
    {
        checked
        {
            while (e.MoveNext())
            {
                count++;
            }
        }
    }

    return count;
}
```

See https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Count.cs

I'm not really happy with that, my implementation would rather look something like this:

```csharp
public static int Count<T>(this IEnumerable<T> enumerable) =>
    enumerable switch
    {
        ICollection<T> collectionOfT => collectionOfT.Count,
        IReadOnlyCollection<T> readOnlyCollectionOfT => readOnlyCollectionOfT.Count,
        string @string => @string.Length,
        ICollection collection => collection.Count,
        _ => DetermineCountViaEnumerating(enumerable)
    };
```

This way, all your corner cases would be covered.

Reply:
There are several reasons why Linq does not check for `IReadOnlyCollection<T>`, performance overhead being one of them (see https://github.com/dotnet/runtime/issues/42254).

I agree that your implementation would cover (almost) all corner cases.

There is only one pathological scenario I can think of where it would enumerate an `ICollection<T>`:
```csharp
public class Base {}

public class Derived : Base {}

public class MyCollection : ICollection<Derived>
{
    // ...
}

public void CornerCase()
{
    // This cast is valid because `IEnumerable<T>` is covariant in its type parameter.
    IEnumerable<Base> enumerable = new MyCollection();
    // Enumerates, as `enumerable is ICollection<Base>` does not hold (`ICollection<T>` is *not* covariant).
    var count = EnumerableExtensions.Count(enumerable);
}
```

IMO this case is pretty far-fetched and can safely be ignored. Note that this would not be an issue if `MyCollection` would additionally implement `IReadOnlyCollection<Derived>`, because `IReadOnlyCollection<T>` *is* covariant, so `enumerable is IReadOnlyCollection<Base>` *would* hold.

Reply:
This morning, I created some benchmarks so that we have data to act upon. The benchmark code looks like this:

```csharp
public class ListCountBenchmark
{
    public List<int> List;
    
    // ReSharper disable once UnusedAutoPropertyAccessor.Global
    [Params(100, 1000)]
    public int NumberOfItems { get; set; }

    [GlobalSetup]
    public void Setup() => List = Enumerable.Range(1, NumberOfItems).ToList();

    // ReSharper disable once UseCollectionCountProperty
    [Benchmark(Baseline = true)]
    public int LinqCount() => List.Count();

    [Benchmark]
    public int ExistingEnumerableCount() => EnumerableExtensions.Count(List);

    [Benchmark]
    public int NewEnumerableOfTGetCount() => List.GetCount();
}

public class StringCountBenchmark
{
    public string String;
    
    // ReSharper disable once UnusedAutoPropertyAccessor.Global
    [Params(100, 1000)]
    public int NumberOfItems { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        var random = new Random(42);
        var array = new char[NumberOfItems];
        for (var i = 0; i < array.Length; i++)
        {
            array[i] = (char) random.Next('a', 'z' + 1);
        }

        String = new (array);
    }

    // ReSharper disable once UseCollectionCountProperty
    [Benchmark(Baseline = true)]
    public int LinqCount() => String.Count();

    [Benchmark]
    public int ExistingEnumerableCount() => EnumerableExtensions.Count(String);

    [Benchmark]
    public int NewEnumerableOfTGetCount() => String.GetCount();
}

public class MyImmutableArrayCountBenchmark
{
    public MyImmutableArray<int> MyArray;
    
    // ReSharper disable once UnusedAutoPropertyAccessor.Global
    [Params(100, 1000)]
    public int NumberOfItems { get; set; }
    
    [GlobalSetup]
    public void Setup() => MyArray = new (Enumerable.Range(1, NumberOfItems).ToArray());
    
    // ReSharper disable once UseCollectionCountProperty
    [Benchmark(Baseline = true)]
    public int LinqCount() => MyArray.Count();

    [Benchmark]
    public int ExistingEnumerableCount() => EnumerableExtensions.Count(MyArray);

    [Benchmark]
    public int NewEnumerableOfTGetCount() => MyArray.GetCount();
}

public sealed class MyImmutableArray<T> : IReadOnlyList<T>
{
    private readonly T[] _array;
    
    public MyImmutableArray(T[] array) => _array = array;

    public IEnumerator<T> GetEnumerator() => ((IEnumerable<T>) _array).GetEnumerator();

    IEnumerator IEnumerable.GetEnumerator() => _array.GetEnumerator();

    public int Count => _array.Length;

    public T this[int index] => _array[index];
}
```

The GetCount() method is implemented as mentioned in my previous comment. We have the following results:

## ListCountBenchmark

```

BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2283/22H2/2022Update/SunValley2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.100-rc.1.23455.8
  [Host]     : .NET 7.0.11 (7.0.1123.42427), X64 RyuJIT AVX2
  Job-RULVGC : .NET 7.0.11 (7.0.1123.42427), X64 RyuJIT AVX2
  Job-EJGZGB : .NET Framework 4.8.1 (4.8.9181.0), X64 RyuJIT VectorSize=256


```
| Method                   | Runtime            | NumberOfItems | Mean      | Error     | StdDev    | Median    | Ratio | RatioSD | Code Size | Allocated | Alloc Ratio |
|------------------------- |------------------- |-------------- |----------:|----------:|----------:|----------:|------:|--------:|----------:|----------:|------------:|
| LinqCount                | .NET 7.0           | 100           | 2.3394 ns | 0.0650 ns | 0.0639 ns | 2.3610 ns |  1.00 |    0.00 |     333 B |         - |          NA |
| ExistingEnumerableCount  | .NET 7.0           | 100           | 0.4343 ns | 0.0086 ns | 0.0081 ns | 0.4401 ns |  0.19 |    0.01 |      46 B |         - |          NA |
| NewEnumerableOfTGetCount | .NET 7.0           | 100           | 3.2079 ns | 0.0224 ns | 0.0209 ns | 3.2040 ns |  1.37 |    0.04 |     199 B |         - |          NA |
|                          |                    |               |           |           |           |           |       |         |           |           |             |
| LinqCount                | .NET Framework 4.8 | 100           | 2.9022 ns | 0.0800 ns | 0.1095 ns | 2.9684 ns |  1.00 |    0.00 |     387 B |         - |          NA |
| ExistingEnumerableCount  | .NET Framework 4.8 | 100           | 4.5974 ns | 0.1081 ns | 0.1011 ns | 4.6108 ns |  1.59 |    0.07 |     105 B |         - |          NA |
| NewEnumerableOfTGetCount | .NET Framework 4.8 | 100           | 3.8110 ns | 0.0110 ns | 0.0103 ns | 3.8093 ns |  1.32 |    0.05 |     217 B |         - |          NA |
|                          |                    |               |           |           |           |           |       |         |           |           |             |
| LinqCount                | .NET 7.0           | 1000          | 2.1248 ns | 0.0122 ns | 0.0114 ns | 2.1205 ns |  1.00 |    0.00 |     333 B |         - |          NA |
| ExistingEnumerableCount  | .NET 7.0           | 1000          | 0.4380 ns | 0.0066 ns | 0.0062 ns | 0.4403 ns |  0.21 |    0.00 |      46 B |         - |          NA |
| NewEnumerableOfTGetCount | .NET 7.0           | 1000          | 3.2074 ns | 0.0202 ns | 0.0189 ns | 3.2155 ns |  1.51 |    0.01 |     199 B |         - |          NA |
|                          |                    |               |           |           |           |           |       |         |           |           |             |
| LinqCount                | .NET Framework 4.8 | 1000          | 2.7832 ns | 0.0088 ns | 0.0073 ns | 2.7813 ns |  1.00 |    0.00 |     387 B |         - |          NA |
| ExistingEnumerableCount  | .NET Framework 4.8 | 1000          | 4.5930 ns | 0.1059 ns | 0.0991 ns | 4.6534 ns |  1.65 |    0.03 |     105 B |         - |          NA |
| NewEnumerableOfTGetCount | .NET Framework 4.8 | 1000          | 3.8097 ns | 0.0151 ns | 0.0141 ns | 3.8185 ns |  1.37 |    0.01 |     217 B |         - |          NA |

## StringCountBenchmark

```

BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2283/22H2/2022Update/SunValley2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.100-rc.1.23455.8
  [Host]     : .NET 7.0.11 (7.0.1123.42427), X64 RyuJIT AVX2
  Job-RULVGC : .NET 7.0.11 (7.0.1123.42427), X64 RyuJIT AVX2
  Job-EJGZGB : .NET Framework 4.8.1 (4.8.9181.0), X64 RyuJIT VectorSize=256


```
| Method                   | Runtime            | NumberOfItems | Mean          | Error      | StdDev     | Median        | Ratio | Code Size | Gen0   | Allocated | Alloc Ratio |
|------------------------- |------------------- |-------------- |--------------:|-----------:|-----------:|--------------:|------:|----------:|-------:|----------:|------------:|
| LinqCount                | .NET 7.0           | 100           |   194.0122 ns |  0.6453 ns |  0.6036 ns |   194.0583 ns | 1.000 |     327 B | 0.0019 |      32 B |        1.00 |
| ExistingEnumerableCount  | .NET 7.0           | 100           |     0.0117 ns |  0.0065 ns |  0.0061 ns |     0.0147 ns | 0.000 |      22 B |      - |         - |        0.00 |
| NewEnumerableOfTGetCount | .NET 7.0           | 100           |     4.9274 ns |  0.0344 ns |  0.0321 ns |     4.9341 ns | 0.025 |     199 B |      - |         - |        0.00 |
|                          |                    |               |               |            |            |               |       |           |        |           |             |
| LinqCount                | .NET Framework 4.8 | 100           |   216.5114 ns |  2.1297 ns |  1.9921 ns |   216.7713 ns |  1.00 |     387 B | 0.0050 |      32 B |        1.00 |
| ExistingEnumerableCount  | .NET Framework 4.8 | 100           |     4.0643 ns |  0.1012 ns |  0.0946 ns |     4.0272 ns |  0.02 |     105 B |      - |         - |        0.00 |
| NewEnumerableOfTGetCount | .NET Framework 4.8 | 100           |    60.6263 ns |  0.9837 ns |  0.9202 ns |    60.4888 ns |  0.28 |     217 B |      - |         - |        0.00 |
|                          |                    |               |               |            |            |               |       |           |        |           |             |
| LinqCount                | .NET 7.0           | 1000          | 1,751.9547 ns |  6.2186 ns |  5.8169 ns | 1,753.7884 ns | 1.000 |     327 B | 0.0019 |      32 B |        1.00 |
| ExistingEnumerableCount  | .NET 7.0           | 1000          |     0.0149 ns |  0.0073 ns |  0.0069 ns |     0.0169 ns | 0.000 |      22 B |      - |         - |        0.00 |
| NewEnumerableOfTGetCount | .NET 7.0           | 1000          |     4.9188 ns |  0.0264 ns |  0.0234 ns |     4.9117 ns | 0.003 |     199 B |      - |         - |        0.00 |
|                          |                    |               |               |            |            |               |       |           |        |           |             |
| LinqCount                | .NET Framework 4.8 | 1000          | 1,810.1125 ns | 35.3886 ns | 42.1276 ns | 1,786.8172 ns | 1.000 |     387 B | 0.0038 |      32 B |        1.00 |
| ExistingEnumerableCount  | .NET Framework 4.8 | 1000          |     4.2021 ns |  0.1026 ns |  0.1008 ns |     4.1399 ns | 0.002 |     105 B |      - |         - |        0.00 |
| NewEnumerableOfTGetCount | .NET Framework 4.8 | 1000          |    59.9531 ns |  0.4154 ns |  0.3886 ns |    59.8529 ns | 0.033 |     217 B |      - |         - |        0.00 |

## MyImmutableArrayBenchmark

```

BenchmarkDotNet v0.13.9+228a464e8be6c580ad9408e98f18813f6407fb5a, Windows 11 (10.0.22621.2283/22H2/2022Update/SunValley2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.100-rc.1.23455.8
  [Host]     : .NET 7.0.11 (7.0.1123.42427), X64 RyuJIT AVX2
  Job-RULVGC : .NET 7.0.11 (7.0.1123.42427), X64 RyuJIT AVX2
  Job-EJGZGB : .NET Framework 4.8.1 (4.8.9181.0), X64 RyuJIT VectorSize=256


```
| Method                   | Runtime            | NumberOfItems | Mean         | Error      | StdDev      | Median       | Ratio | RatioSD | Code Size | Gen0   | Allocated | Alloc Ratio |
|------------------------- |------------------- |-------------- |-------------:|-----------:|------------:|-------------:|------:|--------:|----------:|-------:|----------:|------------:|
| LinqCount                | .NET 7.0           | 100           |   200.590 ns |  4.0402 ns |   8.6969 ns |   204.655 ns |  1.00 |    0.00 |     333 B | 0.0019 |      32 B |        1.00 |
| ExistingEnumerableCount  | .NET 7.0           | 100           |   179.149 ns |  3.4953 ns |   3.7400 ns |   181.121 ns |  0.89 |    0.04 |     363 B | 0.0019 |      32 B |        1.00 |
| NewEnumerableOfTGetCount | .NET 7.0           | 100           |     5.597 ns |  0.0065 ns |   0.0057 ns |     5.598 ns |  0.03 |    0.00 |     199 B |      - |         - |        0.00 |
|                          |                    |               |              |            |             |              |       |         |           |        |           |             |
| LinqCount                | .NET Framework 4.8 | 100           |   192.699 ns |  3.8364 ns |   7.9227 ns |   192.598 ns |  1.00 |    0.00 |     387 B | 0.0050 |      32 B |        1.00 |
| ExistingEnumerableCount  | .NET Framework 4.8 | 100           |   181.953 ns |  3.6389 ns |   8.7885 ns |   186.222 ns |  0.95 |    0.06 |     221 B | 0.0050 |      32 B |        1.00 |
| NewEnumerableOfTGetCount | .NET Framework 4.8 | 100           |    42.715 ns |  0.0961 ns |   0.0852 ns |    42.680 ns |  0.22 |    0.01 |     217 B |      - |         - |        0.00 |
|                          |                    |               |              |            |             |              |       |         |           |        |           |             |
| LinqCount                | .NET 7.0           | 1000          | 1,847.716 ns | 36.8837 ns |  83.2526 ns | 1,879.095 ns | 1.000 |    0.00 |     333 B | 0.0019 |      32 B |        1.00 |
| ExistingEnumerableCount  | .NET 7.0           | 1000          | 1,700.682 ns | 21.6594 ns |  20.2602 ns | 1,707.483 ns | 0.916 |    0.04 |     363 B | 0.0019 |      32 B |        1.00 |
| NewEnumerableOfTGetCount | .NET 7.0           | 1000          |     5.597 ns |  0.0106 ns |   0.0089 ns |     5.599 ns | 0.003 |    0.00 |     199 B |      - |         - |        0.00 |
|                          |                    |               |              |            |             |              |       |         |           |        |           |             |
| LinqCount                | .NET Framework 4.8 | 1000          | 1,729.928 ns | 34.5968 ns |  61.4957 ns | 1,729.409 ns |  1.00 |    0.00 |     387 B | 0.0038 |      32 B |        1.00 |
| ExistingEnumerableCount  | .NET Framework 4.8 | 1000          | 1,524.422 ns | 45.8164 ns | 135.0907 ns | 1,515.641 ns |  0.88 |    0.08 |     221 B | 0.0038 |      32 B |        1.00 |
| NewEnumerableOfTGetCount | .NET Framework 4.8 | 1000          |    42.756 ns |  0.3562 ns |   0.3332 ns |    42.666 ns |  0.02 |    0.00 |     217 B |      - |         - |        0.00 |

## Discussion

I chose the three types for the performance benchmarks because of these reasons:

- `List<T>` should represent the class of common BCL collection types which implements the common collection interfaces. They can be easily cast to `ICollection` and then the `Count` can be accessed. The interesting thing here is that in .NET 7, my existing `Count(this IEnumerable<T> enumerable)` method is way faster than the LINQ `Count` version, where in .NET Framework, it's the other way round.
- `string` was chosen because it is a very common BCL type which is not really supported by LINQ (when it comes to performance optimization). We can see that LINQ's `Count` method will always iterate it, because string does not implement the `ICollection`, `ICollection<T>`, or `IReadOnlyCollection<T>` interfaces. The `Count` implementation of Light.GuardClauses would catch that case (although there's also a huge performance difference between `Count()` and `GetCount()` - maybe that's related to using the switch expression).
- The `MyImmutableArray<T>` becnhmark represents a custom collection which only implements `IReadOnlyList<T>` - both LINQ's and Light.GuardClauses' `Count` methods iterate over it.

## My Conclusion

Regarding these results, I personally wouldn't add the `GetCount` method to Light.GuardClauses - at least not in a v10.x feature release. Maybe for v11 which can contain breaking changes - we could then rename the existing `Count` method to `GetCount` and add the overload for `IEnumerable<T>` (even if it is considerably slower - we should check if we can optimize that). This would be beneficial as we wouldn't have a name clash with the LINQ extension method.

The important point is: when you write custom collections, you need to be aware that you should at least implement `ICollection` so that third-party libraries can efficiently access the count via an abstraction.

But @reima, now I'm interested in your opinion regarding all this data - how would you progress?

Reply:
Thank you for taking the time and creating the benchmarks. It's always valuable to have actual measurements when discussing performance.

I wonder if reordering the switch cases in `GetCount` would be beneficial. Checking for the most common cases first (i.e. `string` and `ICollection`) would trim it down to 1 or 2 type test for these. Besides, I believe type tests for generic types are more expensive, especially for covariant type parameters (like `IReadOnlyCollection<T>`). So doing these last could have an additional benefit for the common case. This is definitely something we should check when considering adding `GetCount`.

The collection sizes you ran the benchmarks with are relatively small. I wonder how the measurements change when considering larger collections (maybe up to 1,000,000 items?), where enumerating them gets considerably more time consuming.

I'm fine with keeping and using the existing `Count` method for now and considering `GetCount` for the next major release. However I would prefer it if the `Range.For` methods would clearly state in its documentation that the enumerable must either be a `string` or implement `ICollection` to get the "fast path".

Reply:
I have now optimized the `GetCount` method - it now looks like this:

```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
[ContractAnnotation("enumerable:null => halt")]
public static int GetCount<T>(this IEnumerable<T> enumerable)
{
    if (enumerable is ICollection collection)
        return collection.Count;
    if (enumerable is string @string)
        return @string.Length;
    if (TryGetCollectionOfTCount(enumerable, out var count))
        return count;

    return DetermineCountViaEnumerating(enumerable);
}

private static bool TryGetCollectionOfTCount<T>([NoEnumeration] this IEnumerable<T> enumerable, out int count)
{
    if (enumerable is ICollection<T> collectionOfT)
    {
        count = collectionOfT.Count;
        return true;
    }
    if (enumerable is IReadOnlyCollection<T> readOnlyCollection)
    {
        count = readOnlyCollection.Count;
        return true;
    }

    count = 0;
    return false;
}
```

The improvements look like this:

| Scenario | Runtime | **Old GetCount** | New GetCount | Count | LINQ Count | **Change** |
| --------- | --------- | -------------------- | ---------------- |---------------------- |-------------- |---------- |
| List<T> 100 Items | .NET 7.0 |  **3.2079 ns** | **1.2999 ns** | 0.4343 ns | 2.3394 ns | **2.47x** ↗️ |
| List<T> 100 Items | .NET Framework 4.8 |  **3.8110 ns** | **4.5462 ns** | 4.5974 ns | 2.9022 ns | **0.84x** ↘️ |
| MyImmutableArray<T> 100 Items | .NET 7.0 | **5.597 ns** | **6.047 ns** | 179.944 ns | 200.590 ns | **0.93x** ↘️ |
| MyImmutableArray<T> 100 Items | .NET Framework 4.8 | **42.715 ns** | **46.168 ns** | 181.953 ns | 192.699 ns | **0.93x** ↘️ |
| String | .NET 7.0 | **4.9274 ns** | **0.2312 ns** | 0.0117 ns | 194.0122 ns | **21.31x** ↗️ |
| String | .NET Framework 4.8 | **60.6263 ns** | **3.9956 ns** | 4.0643 ns | 216.5114 ns | **15.17x** ↗️ |

The table contains the mean times for `GetCount` in its old and new implementation, as well as the mean times for the existing `Count` methods of Light.GuardClauses and LINQ. The last column contains the performance improvement or decline when you divide Old GetCount by New GetCount.

As you can see, we have slight regressions for the `ReadOnlyCollection<T>` cases, as well as for `List<T>` in .NET Framework. The improvements are way greater, from 2.47x up to 21.31x. I will introduce a new `Range.For` overload for `IEnumerable<T>` and call the optimized `GetCount` method internally.

Reply:
I also tested the .NET 7 version under Linux (Ubuntu 22.04 running on WSL 2), the numbers were pretty much the same, so I won't list them here.

Reply:
That's a great performance improvement, thank you for checking! The regression with `MyImmutableArray`/`IReadOnlyCollection<T>` was to be expected. But it's definitely worth it considering the improvents for the more common cases.

I'm puzzled by the slowdown for `List<T>` in .NET Framework. Could be due to the change from the `switch` expression to a sequence of `if`s.